fix(ui): planet labels don't overlap planets (if possible)

flowers9 opened the pull request:
**Bugfix:** This PR addresses issue #8862

## Fix Details
No longer zoom zoom-invariant positional aspects of label positioning when checking for overlaps.  Also, the label is now treated as a full rectangle which is then checked for overlap with system objects, rather than checking three points related to that rectangle, meaning all overlaps should be detected, and it checks for overlap with other labels.  The alpha calculation for fading the label as it leaves the center of the screen has also been corrected so as to be zoom invariant.

Label drawing in Engine has been refactored slightly to speed up drawing labels by calculating positions and overlaps once when entering a system.

Labels now have 8 possible positions, with the new ones above and below the old ones.

Rectangle has a new method to detect overlaps with circles.

## Testing Done
Checking Tarazed on date 17 1 3025, and then going forward day by day until 6 2 3025 (which has a forced overlap when zoomed out).

I checked the generated rectangle by drawing it in with FillShader to see how well it matched the drawn label.

## Save File
[Ron Post.txt](https://github.com/endless-sky/endless-sky/files/12339935/Ron.Post.txt)

## Screenshots
Before:
![Screenshot from 2023-08-14 15-13-08](https://github.com/endless-sky/endless-sky/assets/11161996/34421add-e2db-4f51-9c23-4db4b2a66e56)
After:
![Screenshot from 2023-08-22 17-51-18](https://github.com/endless-sky/endless-sky/assets/11161996/23ee1f4a-4853-4c53-af28-53647fbd6cb7)

Possible Label Positions Before:
![Screenshot from 2023-08-22 15-10-49](https://github.com/endless-sky/endless-sky/assets/11161996/a68c2999-c88f-4428-9cbc-c4375df04d80)
Possible Label Positions After:
![Screenshot from 2023-08-22 15-09-36](https://github.com/endless-sky/endless-sky/assets/11161996/7b52b673-4393-42de-8138-79df4f9c623d)

waterhouse commented:
I tested some of the additional dates from my report.
<img width="1512" alt="Screenshot 2023-08-14 at 6 29 42 PM" src="https://github.com/endless-sky/endless-sky/assets/160180/229b2fa5-850f-4e79-a23f-d102bc1cdcb2">
<img width="1512" alt="Screenshot 2023-08-14 at 6 30 48 PM" src="https://github.com/endless-sky/endless-sky/assets/160180/9e9ff188-9e45-41e0-a082-5e046238a04e">

Definitely an improvement, although I suspect the Echo label is ending up without a good place for it to be, out of the four directions the code chooses from.  Straight up and straight down would work there.  Is there a reason we avoid the 180 and 0 degree offsets?

For completeness:
<img width="1512" alt="Screenshot 2023-08-14 at 6 35 11 PM" src="https://github.com/endless-sky/endless-sky/assets/160180/58cce368-db24-41de-be62-e7430425edcc">
<img width="1512" alt="Screenshot 2023-08-14 at 6 34 43 PM" src="https://github.com/endless-sky/endless-sky/assets/160180/d883dc2b-47db-4db9-8bb5-04ce4564876e">

Again, definitely an improvement, though as you mention it doesn't prevent label overlaps.

flowers9 commented:
Yeah, Echo on 7 2 3025 is forced - there's no non-overlap position, of the four choices.

We could increase the number of positions.  I think the four chosen ones are there because they look nice.  Straight up or down, eh.  There's a preference order, though, so possibly we could just stick 'em in the back, as a last resort.  (Up right is the first resort. ;)

Hurleveur commented:
I think this is an improvement but we could make it even better? There still seems to be some overlapping that we could prevent between the labels themselves?
Could be for another PR tho

flowers9 commented:
Yeah, it turns out to be a pain because we're checking different zoom levels (because they're parts of the zoom placement that are affected by zoom, and parts that aren't).  It'd be pretty easy to do if we were only checking for the current zoom.

flowers9 commented:
Actually, I take that back, it wasn't that bad (just had to store a point and rectangle (which we had already calculated) for each label).  However, it's not necessarily better, because it makes it even more likely to have overlaps in all directions.

Also, it dies horribly near ringworlds.  I'd been thinking the O(n^2) nature of checking labels wouldn't be so bad, because n was always small.  Then I remembered ringworlds, and it gets large enough to be an issue, apparently.  Trying to fly around near one while fully zoomed out is ugly.

Hurleveur commented:
Ring worlds are a bit of an exception and we could probably just not show the same name multiple times

flowers9 commented:
Oh, and I should note that I found and fixed another bug.  When it was calculating the alpha, it was using the zoomed position but it was using the unzoomed object radius.  It seems pretty obvious what the intent was, so I fixed it to match alpha going to zero when the label is far enough away from the center that it's no longer drawn anyway.

flowers9 commented:
And following this to its (more or less) logical conclusion, what do people think of

![Screenshot from 2023-08-23 06-35-45](https://github.com/endless-sky/endless-sky/assets/11161996/dafbe7e2-deb3-48c9-b743-d8dd2ddfbe80)

Note that the attachment points no longer match the original - the incoming line always points to the center of the label bounding box now.  And with that change, any angle can be used, so I tossed in the cardinal directions as well.

flowers9 commented:
Here's an example of how it looks in practice:

![Screenshot from 2023-08-23 09-26-36](https://github.com/endless-sky/endless-sky/assets/11161996/5939c07b-a4e0-4038-83b1-9584c6838b1d)

Zireael07 commented:
@flowers9 Is the line-based solution coded in this PR or have you made pics first?

flowers9 commented:
I've coded it, but I haven't committed it yet.  So, just pics.  I've been keeping the description in the top post up to date with any changes I make to the PR (or I try to, anyway ;).

Since it's a noticeable change, I wanted feedback before adding it.  I suspect people will have opinions.  And it's not really necessary for the purpose of the PR - the 8 position approach should be enough to prevent any reasonable overlap.  But it's nice for allowing arbitrary angles.

Hurleveur commented:
I was mostly mentioning it bc we dont want this PR to make the follow up PR harder than it has to be, and it's good to have both in mind
But yeah we can def split, the more the better for that

Amazinite commented:
Closing this since it's superseded by #9219, and I'm not seeing anything about that PR that would say we wouldn't take it but would take this one.
